### DISCUSSION TITLE
Troubles while using test client and emitting from background thread.

### DISCUSSION THREAD

**Discussion by heitorc62** [miguelgrinberg/Flask-SocketIO#2039] [Q&A] [answered]
Greetings,

I'm encountering a perplexing issue with Flask-SocketIO when attempting to emit from a background thread, specifically in a Dockerized environment utilizing Flask's factory pattern. The core of my project is orchestrated through a run.py file, which calls the create_app function to initialize my Flask application. Initially, all functionalities, including Flask-SocketIO's test client, operated flawlessly. However, the test client ceased to receive emits once I began emitting from a background thread.

To better understand the problem, I've created a simplified version of my project structured as follows:

```
reproduction project
├── /app
└── /tests
```

**Test Code Overview**

Here's a snapshot of the test code employed:
```
import unittest
from app import create_app
from extensions import socketio
import eventlet


class SocketIOTestCase(unittest.TestCase):
    def setUp(self):
        self.app = create_app()
        self.flask_test_client = self.app.test_client()
 

    def tearDown(self):
        pass

    def test_receivement(self):
        # Connect with a valid token
        socketio_test_client = socketio.test_client(self.app, 
                                            flask_test_client=self.flask_test_client)
        
        i = 0
        while (i < 10):
            eventlet.sleep(3)
            r = socketio_test_client.get_received()
            print(f"{i}th response: {r}")
            i += 1
```
            

### Initial Setup
The initial configuration was as follows, situated within the /app directory:
- app.py: Establishes the Flask app, registers a blueprint for SocketIO events, and periodically emits messages in a loop spawned by eventlet.spawn.

--------------app.py---------------
```
import eventlet
eventlet.monkey_patch()

from flask import Flask
from extensions import socketio
import json


def create_app():
    app = Flask(__name__)
    from socketio_events import socketio_bp as socketio_blueprint
    app.register_blueprint(socketio_blueprint)
    socketio.init_app(app, logger=True, engineio_logger=True)
    return app

app = create_app()

def listen():
    while(True):
        socketio.emit('test', dict(foo="bar"), namespace="/")
        eventlet.sleep(2)
        
    
                
eventlet.spawn(listen)


if __name__ == '__main__':
    print("RUNNING!!!!")
    socketio.run(app, host='0.0.0.0', port=5000, use_reloader=False, debug=True, allow_unsafe_werkzeug=True)
```

When executed via Docker compose and unit tests (`docker compose exec flask python -m unittest test.test_socketio`), emissions were successfully received by the test client, as evidenced by the output logs from both the application and the tests.

Look at the output from the application and the test client:
app:
```
test_socketio-flask-1  | Server initialized for eventlet.
test_socketio-flask-1  | RUNNING!!!!
test_socketio-flask-1  | (1) wsgi starting up on http://0.0.0.0:5000
test_socketio-flask-1  | emitting event "test" to all [/]
test_socketio-flask-1  | emitting event "test" to all [/]
test_socketio-flask-1  | emitting event "test" to all [/]
test_socketio-flask-1  | emitting event "test" to all [/]
test_socketio-flask-1  | emitting event "test" to all [/]
test_socketio-flask-1  | emitting event "test" to all [/]
test_socketio-flask-1  | emitting event "test" to all [/]
test_socketio-flask-1  | emitting event "test" to all [/]
test_socketio-flask-1  | emitting event "test" to all [/]
```

test client:
```
Server initialized for eventlet.
Server initialized for eventlet.
Socket.IO client connected: rx1TXeUHFuujxecuAAAA
emitting event "test" to all [/]
emitting event "test" to all [/]
0th response: [{'name': 'test', 'args': [{'foo': 'bar'}], 'namespace': '/'}, {'name': 'test', 'args': [{'foo': 'bar'}], 'namespace': '/'}]
emitting event "test" to all [/]
1th response: [{'name': 'test', 'args': [{'foo': 'bar'}], 'namespace': '/'}]
emitting event "test" to all [/]
emitting event "test" to all [/]
2th response: [{'name': 'test', 'args': [{'foo': 'bar'}], 'namespace': '/'}, {'name': 'test', 'args': [{'foo': 'bar'}], 'namespace': '/'}]
emitting event "test" to all [/]
```

### Modified Setup

I then refactored my project to align more closely with my main project's structure. This involved:
- run.py: Similar functionality as before but separated from app.py to more closely mirror the factory pattern.
- app.py: Modified to only define and return the Flask app.

---------run.py------------
```
import eventlet
eventlet.monkey_patch()

from app import create_app
from extensions import socketio

app = create_app()


def listen():
    while(True):
        socketio.emit('test', dict(foo="bar"), namespace="/")
        eventlet.sleep(2)
        
    
                
eventlet.spawn(listen)


if __name__ == '__main__':
    print("RUNNING!!!!")
    socketio.run(app, host='0.0.0.0', port=5000, use_reloader=False, debug=True, allow_unsafe_werkzeug=True)
```

And modified the `app.py`:
-----------app.py-----------
```
from flask import Flask
from extensions import socketio


def create_app():
    app = Flask(__name__)
    from socketio_events import socketio_bp as socketio_blueprint
    app.register_blueprint(socketio_blueprint)
    socketio.init_app(app, logger=True, engineio_logger=True)
    return app

```

After these adjustments, the test client failed to receive any emits, despite the application logs indicating successful emission.

Look at the output from the application and the test client:
app:
```
test_socketio-flask-1  | Server initialized for eventlet.
test_socketio-flask-1  | RUNNING!!!!
test_socketio-flask-1  | (1) wsgi starting up on http://0.0.0.0:5000
test_socketio-flask-1  | emitting event "test" to all [/]
test_socketio-flask-1  | emitting event "test" to all [/]
test_socketio-flask-1  | emitting event "test" to all [/]
test_socketio-flask-1  | emitting event "test" to all [/]
test_socketio-flask-1  | emitting event "test" to all [/]
test_socketio-flask-1  | emitting event "test" to all [/]
test_socketio-flask-1  | emitting event "test" to all [/]
```

test client:
```
Server initialized for eventlet.
Socket.IO client connected: r7-A_icgFNFFAhA1AAAA
0th response: []
1th response: []
2th response: []
3th response: []
4th response: []
5th response: []
6th response: []
7th response: []
8th response: []
```

## Seeking Insights

I'm puzzled by this discrepancy between the two setups. The only significant change was refactoring the code structure without altering the core logic.

For those interested, my repository contains both setups on the "working-version" branch, allowing for direct comparison:
[First setup (working)](https://github.com/heitorc62/test_socketio/tree/eb2f74e5797a2c206583ef30969e8a585b911a20) 

The second setup can be found at:
[Second setup (not working)](https://github.com/heitorc62/test_socketio/tree/98947faf3121e3ecced03bb4bff12d3a145cb797)

Any insights or suggestions would be greatly appreciated. Thank you for your time and assistance.

**Comment by miguelgrinberg** [comment, marked as answer]
You are using the Flask-SocketIO test client, which is not a real client, it runs directly in the server's context. The case that you think is working is actually running a second server in your client's container, and your test client is actually connecting to that server. You can see the server emitting in the log of your working client above. In the non-working case your client is running but the server is not, so the client connects and does not receive anything.

If you want to run a real client-server test then you can't use the Flask-SocketIO test client, in that case you need to use a real Socket.IO client that connects over the network.

**Reply by heitorc62** [reply]
Thanks for the reply!

> If you want to run a real client-server test then you can't use the Flask-SocketIO test client, in that case you need to use a real Socket.IO client that connects over the network.

Could you suggest a method for executing such a test? Alternatively, could you point me towards an example that illustrates this approach?

**Reply by miguelgrinberg** [reply]
I don't understand what you are asking. The alternative approach I'm suggesting is to use a client to connect to the server. You had a misunderstanding, you thought that the test client is a real client, but it isn't. If you want to use a real cllient, then that's what you should do. The python-socketio repository has client and server examples.
